### DISCUSSION TITLE
Environment variables don't work in Liquidsoap

### DISCUSSION THREAD

**Discussion by juanamm** [savonet/liquidsoap#3348] [Q&A] [answered]
I've spent all night trying to get environment variables to take their value in Liquidsoap.

The result they return is always empty.
`log("Variable SOURCE_PASSW: " ^ getenv("SOURCE_PASSW"))
log("Variable PASSWORD_DJ: " ^ getenv("PASSWORD_DJ"))
log("Variable PASSWORD_ADMIN: " ^ getenv("PASSWORD_ADMIN"))

2023/08/22 09:40:25 [lang:3] Variable SOURCE_PASSW: 
2023/08/22 09:40:25 [lang:3] Variable PASSWORD_DJ: 
2023/08/22 09:40:25 [lang:3] Variable PASSWORD_ADMIN: `

I have defined the variables by editing the `/etc/environment` file so that they are global and always accessible in any environment.

```
PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
TELNET_USERNAME=client_metadata
TELNET_PASSWORD=985421
ADMIN_ICECAST=admin
PASSWORD_ADMIN=9632541
PASSWORD_DJ=20548798
SOURCE_PASSW=05897445
RELAY_PASSW=635478
```
In Ubuntu 22.04 SSH console they work fine:
```
icecast2@ip-17x-xx-xx-xxx:~/liquidsoap-daemon/script$ echo $PASSWORD_ADMIN

9632541
```
Icecast version: Icecast 2.4.4

Liquidsoap version: Liquidsoap 2.1.4 (installed with OCaml Package Manager (opam))

```
icecast2@ip-17x-xx-xx-xxx:~/liquidsoap-daemon$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy
```

Does anyone know a way to make my environment variables work in the .liq script?

**Comment by juanamm** [comment, marked as answer]
**You can close this question**, I already found the solution.
I forgot to mention that I use liquidsoap as a daemon.
I edited the service file and added the line
```
[Service]
EnvironmentFile=/etc/environment
```

And that did the magic. :)

**Comment by vitoyucepi** [comment]
Hi @juanamm,
As you already noticed, common environment variables don't exist inside the systemd service.

There are several ways to load the variables
- `EnvironmentFile`
- `Environment`

You've already used the `EnvironmentFile`, but it's also possible to specify all variables inside the systemd unit file.
```service
[Service]
Environment=TELNET_USERNAME=client_metadata
Environment=TELNET_PASSWORD=985421
```

Storing all variables in the `/etc/environment` is not necessary. Any other file can be used instead.
Consider a script similar to this
1. Create an unprivileged user.
2. Install liquidsoap in user scope.
3. Place the environment file in a user-accessible directory.
4. Create and launch a systemd unit for that user.
5. Enable automatic start of user services at boot.

By the way, it is not necessary to use environment variables; you can load them from a configuration file.
